Let Starlark tests inherit env variables

fmeum opened the pull request:
Adds an inherited_environment field to testing.TestEnvironment to allow
Starlark rules to implement the equivalent of the native rules' 
`env_inherit` attribute.

Work towards #7364. To fully resolve that issue, it remains to handle
executable non-test rules.

RELNOTES: Starlark test rules can use the new inherited_environment
parameter of testing.TestEnvironment to specify environment variables
whose values should be inherited from the shell environment.

fmeum commented:
@meteorcloudy The tests currently fail due to an existing inconsistency: `ActionEnvironment` [gives precedence to inherited variables](https://github.com/bazelbuild/bazel/blob/b3baae9720653a9c6fa8d97780d59331bd590e23/src/main/java/com/google/devtools/build/lib/actions/ActionEnvironment.java#L225), while `env_inherit` on native rules [gives precedence to fixed variables](https://github.com/bazelbuild/bazel/blob/b3baae9720653a9c6fa8d97780d59331bd590e23/src/main/java/com/google/devtools/build/lib/analysis/RunfilesSupport.java#L499).

I actually think that the behavior of `ActionEnvironment` is more useful: Rules could still guard against it by filtering the inherited env list, but can use it to provide sensible defaults. What do you think?

meteorcloudy commented:
@comius might be the right person to answer your question ;)

fmeum commented:
I pushed a new commit that should make the tests pass. I'm now documenting the exact interaction in Starlark (using the `ActionEnvironment` behavior). It is still possible to implement the `env_inherit` behavior on top in case that is desired for starlarkified native rules, so we don't necessarily have to resolve this conflict at this point.

fmeum commented:
The Windows failure has been fixed.

meteorcloudy reviewed (commented):
Thanks, it generally looks good to me, but @comius should do the final review.

fmeum commented:
@comius Friendly ping. Can you already say roughly when you would be able to review this?

comius commented:
> Work towards https://github.com/bazelbuild/bazel/issues/7364

What else is coming? Having a plan first might help making things more consistent.

comius commented on src/main/java/com/google/devtools/build/lib/starlarkbuildapi/test/TestEnvironmentInfoApi.java:
```diff
@@ -28,4 +29,10 @@
+
+  @StarlarkMethod(
+      name = "inherited_environment",
+      doc = "A list of variables that the test action should inherit from the environment.",
```
Can you make this more precise? From which "environment" are they inherited?

fmeum replied:
I added "shell" to clarify this.

comius commented on src/main/java/com/google/devtools/build/lib/actions/ActionEnvironment.java:
```diff
@@ -186,6 +186,23 @@ public ActionEnvironment addFixedVariables(Map<String, String> vars) {
+    if (inheritedVars.isEmpty()) {
+      return addFixedVariables(vars);
+    } else {
+      // TODO: inheritedEnv is currently not optimized for allocation avoidance in the same way as
+      //  fixedEnv.
```
Do you plan to address this? It might become an issue.

fmeum replied:
I will look into doing it now, but could also reserve it for a follow-up PR. Let me know what you prefer.

fmeum replied:
I pushed another that implements the optimization.

comius commented on src/main/java/com/google/devtools/build/lib/starlarkbuildapi/test/TestingModuleApi.java:
```diff
@@ -56,12 +58,24 @@ ExecutionInfoApi executionInfo(Dict<?, ?> requirements // <String, String> expec
+              named = true,
+              positional = true,
+              doc =
+                  "A sequence of string keys that represent environment variables. These variables"
+                      + " will be made available during the test execution with their current value"
+                      + " taken from the environment. If a variable is contained in both"
+                      + " <code>environment</code> and <code>inherited_environment</code>, the"
+                      + " value inherited from the environment will take precedence if set.")
```
> If a variable is contained in both <code>environment</code> and <code>inherited_environment</code>, the value inherited from the environment will take precedence if set.

This is the opposite of implementation in ActionEnvironment. I would also think it makes more sense that env takes precedence.

fmeum replied:
Thanks, I modified the comment as you suggested.

> This is the opposite of implementation in ActionEnvironment. I would also think it makes more sense that env takes precedence.

Sorry, I think the missing `shell` qualifier in front of environment makes this comment confusing. Updated it to say "the value inherited from the **shell** environment will take precedence if set", which should align with the behavior of `ActionEnvironment#resolve`.
Is this the behavior you find most reasonable? I do, simply because it allows rules to choose the behavior they want: They can provide an overridable default by specifying a variable name in both `environment` and `inherited_environment`, but they can also ensure that the fixed value in `environment` is always used simply by omitting it from `inherited_environment`. If `environment` always were to take precedence, the former behavior wouldn't be possible to achieve. 

Note though that the current behavior of `env` and `env_inherit` on native rules is in fact different. [Here, `env` always takes precendence](https://github.com/bazelbuild/bazel/blob/bd35a08674fafdccf71dcc525faa2aa09824bfa7/src/main/java/com/google/devtools/build/lib/analysis/RunfilesSupport.java#L499). But since the new provider field would still allow to implement the current behavior, that shouldn't cause problems for starlarkification.

fmeum replied:
I just noticed that `ActionEnvironment` says "The order in which the environments are combined is undefined, so callers need to take care that the key set of the {@code fixedEnv} map and the set of {@code inheritedEnv} elements are disjoint.". Is there a reason not to specify one or the other behavior?

comius replied:
> Is there a reason not to specify one or the other behavior?

I think the following options are valid:
- define the behavior 
- throw an error when sets are not disjoint

Saying it's undefined and asking the user to keep it disjoint but not enforcing it is the worse choice. So feel free to update it.

> @comius Friendly ping. If you could clarify your preference regarding the fixedEnv/inheritedEnv precedence, I could finalize the PR ahead of your final review.

I agree current order gives you a bit more choice (i.e. having defaults possibly overridden from env) than the opposite (inherit but always override, where duplication doesn't make sense).

fmeum commented:
In the emails I received from GitHub, I also saw this comment

> This should be ImmutableList (and ImmutableMap above). We don't want data inside a provider to mutate.

I can't find it now, but made the suggested change. Let me know if you didn't want me to do that.

fmeum commented:
@comius Friendly ping. If you could clarify your preference regarding the fixedEnv/inheritedEnv precedence, I could finalize the PR ahead of your final review.

comius reviewed (approved):
I only found a couple of nit picks, like the ones mentioned below and using `assertThat(x).containsEntry` where possible. If you don't have any objections, I'll just make those changes myself.

fmeum commented on src/main/java/com/google/devtools/build/lib/actions/ActionEnvironment.java:
```diff
@@ -84,48 +106,66 @@ public boolean isEmpty() {
     @Override
-    public ImmutableMap<String, String> toMap() {
+    public ImmutableMap<String, String> getFixedEnvironment() {
       Map<String, String> result = new LinkedHashMap<>();
```
Done.

fmeum commented on src/main/java/com/google/devtools/build/lib/actions/ActionEnvironment.java:
```diff
@@ -84,48 +106,66 @@ public boolean isEmpty() {
+
+    @Override
+    public ImmutableSet<String> getInheritedEnvironment() {
+      Set<String> result = new LinkedHashSet<>();
```
Done.

fmeum commented:
> I only found a couple of nit picks, like the ones mentioned below and using assertThat(x).containsEntry where possible. If you don't have any objections, I'll just make those changes myself.

I added a commit that uses `Immutable{Map,Set}.builder()` instead as you suggested. Feel free to make any other changes as you see fit.

> Saying it's undefined and asking the user to keep it disjoint but not enforcing it is the worse choice. So feel free to update it.

I added a commit that defines `inheritedEnv` to take precedence over `fixedEnv` and also updated some comments and function names to reflect that.

> This change results in some memory improvements of Bazel:

That's a nice bonus. I also added a commit that should further improve the interning of empty values.

fmeum commented:
@comius If you prefer, I could also split of the last four commits into a separate PR. They don't really have much to do with the original goal of this PR anymore. But if you can just do that when importing the change (or don't see a reason to split it up further), that is also fine with me.

comius commented:
> @comius If you prefer, I could also split of the last four commits into a separate PR. They don't really have much to do with the original goal of this PR anymore. But if you can just do that when importing the change (or don't see a reason to split it up further), that is also fine with me.

Could you reopen a PR with remaining changes after this is merged? git rebase should handle it nicely. I had to make some internal only changes, and I don't have a nice/easy way to combine your current state with them.

fmeum commented:
> > @comius If you prefer, I could also split of the last four commits into a separate PR. They don't really have much to do with the original goal of this PR anymore. But if you can just do that when importing the change (or don't see a reason to split it up further), that is also fine with me.
> 
> Could you reopen a PR with remaining changes after this is merged? git rebase should handle it nicely. I had to make some internal only changes, and I don't have a nice/easy way to combine your current state with them.

Sounds good. Will leave this PR as is, but open a new one with whatever hasn't been merged once this one has been closed by the bot account.

brentleyjones commented:
@bazel-io flag

ckolli5 commented:
@bazel-io fork 5.2.0
